### DISCUSSION TITLE
What to do instead of using deprecated watch() method

### DISCUSSION THREAD

**Discussion by halbekanne** [state-adapt/state-adapt#53] [Q&A] [answered]
I'm just exploring this library and came across something that I solved using `watch()` but seeing as it is deprecated I'm not sure about how to rewrite this.
Probably easiest to explain what I want to do:

I have a chat application where I first fetch chats. This http call also gives me the IDs of participants in these chats.
I collect those IDs and ask the backend for more information about these participants, then add them to the state.

For the part where I have to fetch the backend a second time for more detailed information, I used the following code with the watch method:

```ts
  private fetchParticipantDetailsSource = getRequestSources(
    '[Chat] fetch participant details',
    watch(this.storePath, chatStateAdapter).participantsWithNoDetails$.pipe(
      ...
      switchMap((participantsWithNoDetails) =>
        this.chatHttpService.getUserDetails$(participantsWithNoDetails)
      )
    )
  );
```

My store looks like this:

```ts
  private chatStore = adapt(
    [this.storePath, this.initialState, chatStateAdapter],
    {
      chatsFetched: this.fetchChatsSources.success$,
      chatsFetchedError: this.fetchChatsSources.error$,
      participantDetailsFetched: this.fetchParticipantDetailsSource.success$,
      participantDetailsFetchedError: this.fetchParticipantDetailsSource.error$,
    }
  );
```

This works like intended, but it says in the documentation that using `watch()` is deprecated and I should look into doing it with `adapt()`somehow. This is the part where I need help, how do I rewrite the code to use the recommended method? I have looked over the documentation of that method but I can't understand what I'm supposed to be doing instead of `watch()`.

**Comment by mfp22** [comment, marked as answer]
I think I should remove the deprecation for `watch`. There are some rare but valid scenarios where it's needed and `adapt` itself isn't enough. But the original thought was that `adapt` can now handle these circular things, because sources can be defined with a function:

```typescript 
  private chatStore = adapt(
    [this.storePath, this.initialState, chatStateAdapter],
    store => {
      const fetchParticipantDetailsSources = // use store.participantsWithNoDetails$
      return {
        chatsFetched: this.fetchChatsSources.success$,
        chatsFetchedError: this.fetchChatsSources.error$,
        participantDetailsFetched: fetchParticipantDetailsSources.success$,
        participantDetailsFetchedError: fetchParticipantDetailsSources.error$,
      };
    } 
  );
```

**Reply by halbekanne** [reply]
Thanks for the code example, it helped me understand how I could move forward. I also looked at the books example, it helps with understanding how the stores can be used.
